[Distributed] Re-implement ad-hoc requirements into dynamic witness table lookup for `SerializationRequirement` conformance

xedin opened the pull request:
- `SerializationRequirement` of `DistributedActorSystem` became a primary associated type
- `remoteCall` and `remoteCallVoid` of `DistributedActorSystem` are now requirements
- Sema create abstract conformances for `Res` generic parameter of each `remoteCall` witness
- IRGen is adjusted to produce witness tables for `Res: SerializationRequirement` that is required
  for every witness (but cannot be expressed on the requirement) of `DistributedActorSystem.remoteCall` 
  dynamically in a protocol witness thunk.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintSystem] NFC: Pass an underlying declaration to `Solution::computeSubstitutions`

This is going to be used to determine whether the substitutions are
computed for `DistributedActorSystem::remoteCall` and adjust the
generic signature with witness conformance requirements.

xedin committed:
[stdlib] Distributed: Promote `remoteCall{Void}` into protocol requirements

xedin commented:
This is still a draft because we need to adjust distributed thunk synthesis to not assume that `actorSystem` is always concrete and add some more Sema checks to make sure that `SerializationRequirement` is always supplied to `distributed actor` declarations.

ktoso commented on lib/Sema/TypeCheckProtocol.cpp:
```diff
@@ -1195,13 +1195,52 @@ swift::matchWitness(WitnessChecker::RequirementEnvironmentCache &reqEnvCache,
+        // of `DistributedActorSystem.remoteCall` are not expressible at the moment
+        // but they are verified by Sema so it's okay to omit them here and lookup
+        // dynamically during IRGen.
+        if (funcDecl && funcDecl->isDistributedActorSystemRemoteCallRequirement(
```
Fantastic, thanks for figuring this out @xedin !

xedin committed:
[IRGen] Implement dynamic witness table lookup for `Res` of `DistributedActorSystem.remoteCall` witnesses

When IRGen is building a protocol witness thunk for a
`DistributedActorSystem.remoteCall` requirement we
need to supply witness tables associated with `Res`
generic parameter which are not expressible on the
requirement because they come from `SerializationRequirement`
associated type.

xedin committed:
[TypeChecker] Allow missing requirements on `Res` generic parameter of `DistributedActorSystem.remoteCall`

Conformance requirement between on `Res` and `SerializationRequirement`
of `DistributedActorSystem.remoteCall` are not expressible at the moment
but they are verified by Sema so it's okay to omit them here and lookup
dynamically during IRGen.

xedin committed:
[CSApply] Produce an abstract conformance for `Res` parameter of `DistributedActorSystem.remoteCall` witness

xedin committed:
[CSSimplify] Synthesize conformances for ad-hoc distributed witness requirements

Instead of handing failures in `matchWitness` let's syntehsize
conformances during solving. Next step would be to record them
and use in `Solution::computeSubstitutions`.

xedin committed:
[ConstraintSystem] Record conformances synthesized for ad-hoc distributed requirements

xedin committed:
[Sema/IRGen] Extend ad-hoc requirement handling to `DistributedTargetInvocationEncoder.record{Argument, ReturnType}`

xedin committed:
[Sema/IRGen] Extend ad-hoc requirement handling to `DistributedTargetInvocationDecoder.decodeNextArgument`

xedin committed:
[Sema/IRGen] Extend ad-hoc requirement handling to `DistributedTargetInvocationResultHandler.onReturn`

xedin committed:
[stdlib] Distributed: Remove `invokeOnReturn` requirement and its synthesis

This is no longer necessary because `onReturn` is a protocol
requirement now.

xedin committed:
[SILOptimizer] Distributed: Suppress signature specialization for witnesses with ad-hoc serialization requirement

xedin committed:
[Tests] NFC: Adjust distributed actor test-cases changed/improved due to new requirements

ktoso commented on lib/Sema/DerivedConformanceDistributedActor.cpp:
```diff
@@ -121,318 +121,6 @@ static FuncDecl *deriveDistributedActor_resolve(DerivedConformance &derived) {
-  derived.addMembersToConformanceContext({funcDecl});
-  return funcDecl;
-}
-
```
so much code gone, yaaaaaay! 🥳

ktoso reviewed (approved):
Looking great so far

xedin committed:
[Distributed] Promote SerializationRequirement as a primary associated type for encoder/decoder/result

xedin committed:
[Sema] Distributed: Adjust distributed thunk synthesis to use witnesses for `actorSystem` references

Things like `makeInvocationEncoder()` and all of the `record*`
methods can now be called through witnesses.

xedin committed:
[IRGen] Distributed: Don't attempt to inject protocols without witness tables

xedin committed:
[IRGen] Distributed: Always invoke `decodeNextArgument` through witness thunk

`decodeNextArgument` is now a requirement, so thunks should be
always calling that instead of trying to dispatch directly.

xedin committed:
[SIL] Distributed: Remove logic related to ad-hoc requirements from SILFunction

Ad-hoc requirements are now obsolete by making `remoteCall`,
`record{Argument, ReturnType}`, `decodeNextArgument` protocols
requirements and injecting witness tables for `SerializationRequirement`
conformances during IRGen.

xedin committed:
[AST] Distributed: Make sure that prospective `remoteCall` declaration is in type context

ktoso commented on include/swift/SIL/SILFunction.h:
```diff
@@ -264,17 +264,6 @@ class SILFunction
-  /// 'decodeNextArgument' which must be retained, as it is only used from IRGen
-  /// and such, appears as-if unused in SIL and would get optimized away.
-  // TODO: Consider making this a general "references adhoc functions" and make it an array?
-  SILFunction *RefAdHocRequirementFunction = nullptr;
```
Awesome to see this gone

ktoso commented on lib/IRGen/GenProto.cpp:
```diff
@@ -722,6 +733,63 @@ bool EmitPolymorphicParameters::isClassPointerSource(unsigned paramIndex) {
+void EmitPolymorphicParameters::injectAdHocDistributedRequirements() {
+  // FIXME: We need a better way to recognize that function is
+  // a thunk for witness of `remoteCall` requirement.
+  if (!Fn.hasLocation())
```
In my other PR I was actually going to add:

```
  /// Record which function this is a thunk for, we'll need this to link back
   /// calls in case this is a distributed requirement witness.
   thunk->getAttrs().add(
       new (C) DistributedThunkTargetAttr(func));
```

so we could check for it to detect a thunk; 
https://github.com/apple/swift/pull/70928/files#diff-c5ee3c123c92773a4f07b456c2c0051c7a9dfdc81c5c0234ba5da6389bd82a07R793

I'll cleanup  that PR shortly

xedin replied:
Great! Once that's there we can use it both here and in `FunctionSignatureOpts` optimizer pass.

ktoso reviewed (approved):
Incredible work here, thanks so much @xedin. Great to see all the horrible ad hoc requirements gone.

Read through in depth, LGTM!

xedin committed:
[AST] WitnessMatching: Bring back original check for missing Sendable conformances

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please build toolchain macOS platform
